sqlx-postgres/src/types/cube.rs

```diff
@@ -0,0 +1,363 @@
+                            &bytes,
+                            start_index + step_idx * step + dim_idx * BYTE_WIDTH,
+                        )
+                    })
+                    .collect()
+            })
+            .collect()
+    }
+
+    fn parse_float_from_str(s: &str, error_msg: &str) -> Result<f64, Error> {
+        s.trim()
+            .parse()
+            .map_err(|_| Error::Decode(error_msg.into()))
+    }
+
+    fn parse_point(str: &str) -> Result<PgCube, Error> {
+        Ok(PgCube::Point(Self::parse_float_from_str(
+            str,
+            "Failed to parse point",
+        )?))
+    }
+
+    fn parse_zero_volume(content: &str) -> Result<PgCube, Error> {
+        content
+            .split(',')
+            .map(|p| Self::parse_float_from_str(p, "Failed to parse into zero-volume cube"))
+            .collect::<Result<Vec<_>, _>>()
+            .map(PgCube::ZeroVolume)
+    }
+
+    fn parse_one_dimensional_interval(point_vecs: Vec<&str>) -> Result<PgCube, Error> {
+        let x = Self::parse_float_from_str(
+            &Self::remove_parentheses(point_vecs[0]),
+            "Failed to parse X in one-dimensional interval",
+        )?;
+        let y = Self::parse_float_from_str(
+            &Self::remove_parentheses(point_vecs[1]),
+            "Failed to parse Y in one-dimensional interval",
+        )?;
+        Ok(PgCube::OneDimensionInterval(x, y))
+    }
+
+    fn parse_multidimensional_interval(point_vecs: Vec<&str>) -> Result<PgCube, Error> {
+        point_vecs
+            .iter()
+            .map(|&point_vec| {
+                point_vec
+                    .split(',')
+                    .map(|point| {
+                        Self::parse_float_from_str(
+                            &Self::remove_parentheses(point),
+                            "Failed to parse into multi-dimension cube",
+                        )
+                    })
+                    .collect::<Result<Vec<_>, _>>()
+            })
+            .collect::<Result<Vec<_>, _>>()
+            .map(PgCube::MultiDimension)
+    }
+
+    fn deserialize_string(input: &str) -> Result<PgCube, Error> {
+        let content = &input[1..input.len() - 1];
+
+        if !content.contains('(') && !content.contains(',') {
+            return Self::parse_point(content);
+        }
+
+        if !content.contains("),(") {
+            return Self::parse_zero_volume(content);
+        }
+
+        let point_vecs = content.split("),(").collect::<Vec<&str>>();
+        if point_vecs.len() == 2 && !point_vecs.iter().any(|pv| pv.contains(',')) {
+            return Self::parse_one_dimensional_interval(point_vecs);
+        }
+
+        Self::parse_multidimensional_interval(point_vecs)
+    }
+
+    fn remove_parentheses(s: &str) -> String {
+        s.trim_matches(|c| c == '(' || c == ')').to_string()
+    }
+
+    fn deserialize_bytes(bytes: &[u8]) -> Result<Self, Error> {
+        let cube_type = bytes[0] as usize;
```

Comment:
Again, anything using the `[]` operator will panic if the index is out of bounds. This needs to be carefully checked to avoid surprising panics (e.g. https://github.com/launchbadge/sqlx/issues/2691).

Reply:
I've moved towards the `get` method, and will throw an error with the inputted bytes/string if it returns none 👍